agent/lib_aws_sdk_php.c

```diff
@@ -0,0 +1,177 @@
+ * loaded yet and then evaluate the string. However, in the rare case that files
+ * are not loaded via autoloader and/or have non-default composer classload
+ * settings, if the class is not found, PHP 8.2+ will generate a fatal
+ * unrecoverable uncatchable error error whenever it cannot find a class. While
+ * calling this from nr_aws_sdk_php_enable would have been great and would allow
+ * the sdk version value to be set only once, to avoid the very unlikely but not
+ * impossible fatal error, this will be called from the
+ * "Aws\\ClientResolver::_apply_user_agent" wrapper which GUARANTEES that
+ * aws/sdk exists and is already loaded.
+ *
+ *
+ * Additionally given that aws-sdk-php is currently detected from the
+ * AwsClient.php file, this method will always be called when a client is
+ * created unlike Sdk::construct which doesn't show with PHP 8.2+.
+ *
+ * Using Aws/Sdk::__construct for version is currently nonviable as it is
+ * unreliable as a version determiner.
+ * Having separate functionality to extract from Aws/Sdk::__construct
+ * is both not required and is redundant and causes additional overhead and
+ * so only one function is needed to extract version.
+ *
+ * Aws\\ClientResolver::_apply_user_agent a reliable function as it is
+ * always called on client initialization since it is key to populating
+ * the request headers, and it loads Sdk by default.
+ *
+ * Concerns about future/past proofing to the checking prioritized the following
+ * implementation vs using the eval method.
+ */
+void nr_lib_aws_sdk_php_handle_version() {
+  zval* zval_version = NULL;
+  zend_class_entry* class_entry = NULL;
+  char* version = NULL;
+
+  class_entry = nr_php_find_class("aws\\sdk");
+  if (NULL != class_entry) {
+    zval_version = nr_php_get_class_constant(class_entry, "VERSION");
+
+    if (nr_php_is_zval_non_empty_string(zval_version)) {
+      version = Z_STRVAL_P(zval_version);
+    }
+  }
+  if (NRINI(vulnerability_management_package_detection_enabled)) {
+    /* Add php package to transaction */
+    nr_txn_add_php_package(NRPRG(txn), PHP_PACKAGE_NAME, version);
+  }
+  nr_fw_support_add_package_supportability_metric(NRPRG(txn), PHP_PACKAGE_NAME,
+                                                  version);
+  nr_php_zval_free(&zval_version);
+}
+
+void nr_lib_aws_sdk_php_add_supportability_service_metric(
+    const char* service_name) {
+  int MAX_LEN = 512;
+  char buf[MAX_LEN];
+
+  if (NULL == service_name || '\0' == service_name[0]) {
+    return;
+  }
+  if (NULL == NRPRG(txn)) {
+    return;
+  }
+  buf[0] = '\0';
+
+  snprintf(buf, MAX_LEN, "%s%s", PHP_AWS_SDK_SERVICE_NAME_METRIC_PREFIX,
+           service_name);
+  nrm_force_add(NRPRG(txn) ? NRTXN(unscoped_metrics) : 0, buf, 0);
+}
+
+NR_PHP_WRAPPER(nr_aws_version) {
+  (void)wraprec;
+  NR_PHP_WRAPPER_CALL;
+  nr_lib_aws_sdk_php_handle_version();
+}
+NR_PHP_WRAPPER_END
+
+/*
+ * AwsClient::parseClass
```

Comment:
Leaving this for posterity: https://github.com/aws/aws-sdk-php/blob/3.316.3/src/AwsClient.php#L373-L387